### PR TITLE
Implementation of fine tuning per GPU

### PR DESCRIPTION
**Pull request by AndreaLanfranchi**
This changes allow to fine tune even the slightest differences among
different GPUs and/or help in setting up mixed cards rigs (eg. 2x Gtx
1080 + 4x Gtx 1070)
It also reduces the calls to nvidia-smi and nvidia-settings as much as
possible.
Introduced some cosmetics in the output of nvidia-overclock
Minor changes in monitor pages titles.

### CODE REVIEW

**Comment by invidtiv** [comment]
Great work, anychance of getting to work with 13GPUs ?

**Comment by AndreaLanfranchi** [comment]
@invidtiv 
It should ... logic is implemented up to 99 GPUs

**Comment by invidtiv** [comment]
works very well. Thank you for your work.

**Comment by Cyclenerd** [comment]
Great change. I think the change leaves everything still easy to read and understand. Well done. Already gave the user request for overclocking for individual cards. Will certainly liked!

A few things:

* Please do not disable shellchecks. Please implement all warnings.
* You can also have your scripts tested on this website: https://www.shellcheck.net/
* Please also adjust [SC1117](https://github.com/koalaman/shellcheck/wiki/SC1117) or just use echo: `Backslash is literal in "\n". Prefer explicit escaping: "\\n"`
https://www.shellcheck.net/
* All changes with setup are written to the end of the file `settings.conf` (https://github.com/Cyclenerd/ethereum_nvidia_miner/blob/master/files/setup.sh#L246). So you have to have all the comments above.

**Comment by AndreaLanfranchi** [comment]
Hi, following your points

- Some checks are really cosmetic. Implementing all of them sometimes is not worth it. Please also consider that many checks occur due to the fact that variables like MY_WATT etc are all uppercase (which in shell would mean they are environ variables) while in reality they are always only local variables to scripts (would suggest to change casing)
- Yes but sometimes all copy and paste is quite boring
- Will not use echo due to this https://stackoverflow.com/questions/8467424/echo-newline-in-bash-prints-literal-n. Probably you have fallen in such behavior as i saw a couple of 

```
echo 
echo "Done"
echo

```

to simulate new line.

- Will look into setup.sh

**Comment by AndreaLanfranchi** [comment]
Would also like to ask you what's the point of having Fail2Ban installed. It introduces a great CPU overhead while (I presume) almost noone will have their miner(s) directly exposed to the internet. (aside the fact they may want to remotely check munin).

**Comment by Cyclenerd** [comment]
I think many put the SSH port on the Internet to quickly solve problems. I do it that way too. Fail2Ban is a welcome security gain. Also, I do not run any server without Fail2Ban.

I also had some miners in a RZ in other EU countries (cheaper electricity than in Germany). The network was local but not trustworthy.

I like to use echo because the formatting is in the source code well visible and everyone really understands it.

**Comment by AndreaLanfranchi** [comment]
Do you agree to change casing for local variables ?

**Comment by Cyclenerd** [comment]
D rather not. I'm so used to it :)

I'm just wondering if it makes sense to build a function in nvidia-overclock.sh which you can also call from miner.sh. What do you think?

**Comment by AndreaLanfranchi** [comment]
> I'm just wondering if it makes sense to build a function in nvidia-overclock.sh which you can also call from miner.sh. What do you think?

Not sure I understand what you mean.
You can of course invoke nvidia-overclock.sh from within miner.sh

**Comment by Cyclenerd** [comment]
Exactly. Then the source code for the power limit would not be double.

**Comment by AndreaLanfranchi** [comment]
Ok now I see ... will reason over it. At first glance a parameter check on top will be enough.

**Comment by Cyclenerd** [comment]
May I include your changes as I mean and you look over it. I would then create an extra branch.

**Comment by AndreaLanfranchi** [comment]
Ok.

**Comment by Cyclenerd** [comment]
Done and successfully tested: https://github.com/Cyclenerd/ethereum_nvidia_miner/tree/fine-tuning-per-gpu-

What do you think?

File: files/nvidia-overclock.sh [hunk 1]
```diff
@@ -6,25 +6,85 @@
...
-	# Set Fan speed , Graphics clock, Memory Clock
-	nvidia-settings -a "[gpu:$MY_DEVICE]/GPUPowerMizerMode=1" -a "[gpu:$MY_DEVICE]/GPUFanControlState=1" -a "[fan:$MY_DEVICE]/GPUTargetFanSpeed=$MY_FAN" -a "[gpu:$MY_DEVICE]/GPUGraphicsClockOffset[3]=$MY_CLOCK" -a "[gpu:$MY_DEVICE]/GPUMemoryTransferRateOffset[3]=$MY_MEM"
-	# Set watt/powerlimit. This is also set in miner.sh at autostart.
-	sudo nvidia-smi -i "$MY_DEVICE" -pl "$MY_WATT"
+
+    printf "\n  Processing GPU %s \n  ------------------------------------------------------------------------------\n" "$MY_DEVICE"
+
+    # Init arguments
+    # PowerMizer always se
+    MY_ARG=" -a \"[gpu:$MY_DEVICE]/GPUPowerMizerMode=1\""
```
Comment (invidtiv) [inline files/nvidia-overclock.sh:52]: MY_ARG=" -a \"[gpu:$MY_DEVICE]/GPUPowerMizerMode=1\" -a "[gpu:$MY_DEVICE]/GPUFanControlState=1" ".

Without GPUFanControlState Asus GTX1070 do not accept fan value.

Comment (Cyclenerd) [reply files/nvidia-overclock.sh:52]: Added: https://github.com/Cyclenerd/ethereum_nvidia_miner/commit/34117053db537abd0445513a1f9c040cf6a8ed5d

**Comment by AndreaLanfranchi** [comment]
@invidtiv you're right.
But I'd leave the addition with the fan section so we can set fancontrolstate to 1 IF the user enters a specific fan percentage value, or revert it back to zero (auto) if no specific value

**Comment by AndreaLanfranchi** [comment]
I'll commit my proposal

**Comment by invidtiv** [comment]
@AndreaLanfranchi ,
yes that's even better.
Thank you

File: files/nvidia-overclock.sh (hunk 1 above)
Comment (Cyclenerd) [reply files/nvidia-overclock.sh:52]: Changed: https://github.com/Cyclenerd/ethereum_nvidia_miner/commit/0412c90edf90981e4c4927eaa7b927df3ff50873

Comment (Cyclenerd) [reply files/nvidia-overclock.sh:52]: And: https://github.com/Cyclenerd/ethereum_nvidia_miner/commit/57d2a1601b43df565937c879c36c850e66e6bb9e

**Comment by AndreaLanfranchi** [comment]
> All changes with setup are written to the end of the file settings.conf (https://github.com/Cyclenerd/ethereum_nvidia_miner/blob/master/files/setup.sh#L246). So you have to have all the comments above.

Setup.sh has been modified so each parameter entered is replaced at the same line position within settings.conf

**Comment by bartocc** [comment]
I'll try this nice PR as soon as possible

**Comment by bartocc** [comment]
Just as a side note (this might be a good idea for another PR), I run 1 miner per GPU. This way, if the miner fails for any reason (too overclocked most of the time), I still mine with the others. I've also have the feeling that I get a better accepted shares rate when doing this.

**Comment by Cyclenerd** [comment]
I will close this pull because in the meantime it contains a lot of new features.

The support to overclock single cards is meanwhile in the master (https://github.com/Cyclenerd/ethereum_nvidia_miner/commit/a6dcac668fe68b657b227f01070fa388d360fd2d). 

For further changes and new functions (`setup.sh`) please create a new pull.

**Comment by AndreaLanfranchi** [comment]
@bartocc 
I see several people doing this ... however I highly discourage it.
Of course there may be perceptions which encourage this behavior but, **if you run ethminer**, you should take into account :

- Every instance of ethminer uses at least one thread plus 1 thread per gpu. This means that for a rig of 6 gpus you have 7 threads for a single instance of ethminer while in "single gpu mode" you would end up with 12 threads. CPU (not GPU) stress i higher.
- A single instance of ethminer "clusters" work of gpus on same jobs: ethminer gives each Gpu a non-overlapping range of nonces to search for. I.e. for every job received by the pool you have 6 workers working on different nonces on the same job. On the other hand in "single GPU mode" you end up with every Gpu working on its own job pushed by the pool;
- A single instance of ethminer keeps only 1 socket connected to the pool: in "single GPU mode" you get 6 sockets which lock each other ( as sockets are TCP not UDP)

This all said I do not mean to go agains your perceptions but they must be weighted by objective data.
Best

**Comment by AndreaLanfranchi** [comment]
@bartocc 

> Just as a side note (this might be a good idea for another PR), I run 1 miner per GPU. This way, if the miner fails for any reason (too overclocked most of the time), I still mine with the others. I've also have the feeling that I get a better accepted shares rate when doing this.

You can now fine tune overclocking per single GPU thus your fail occurences due to excessive oc might be mitigated.

**Comment by bartocc** [comment]
thx a lot @AndreaLanfranchi! The second point is particularly interesting.

Fine tuning each of my GPUs is indeed gonna help. The miner.sh process is still likely to fail at one point in time though, and I wish we had some monitoring process to auto-restart it in such cases.

**Comment by AndreaLanfranchi** [comment]
@bartocc 
Fine ... it's not miner.sh wich fails but the ethminer itself (it may happen due to heavy overclocking): in factory values you can run ethminer for months without a glitch.
Will continue to develop on my fork : a watchdog is in priority list.

**Comment by bartocc** [comment]
true, and true ;-) I've been running on factory settings for quite a while.

Overclocking is still pretty interesting though.

Thanks for your work!

**Pull request closed without merging**